File: server/common/utils/crc.go
package utils

func CRC8XOR(data []byte) byte {
	var crc byte
	for _, b := range data {
		crc ^= b
	}

	return crc
}

func CheckCRC(buffer []byte) bool {
	if len(buffer) < 2 {
		return false
	}

	data := buffer[:len(buffer)-1]
	expectedCRC := buffer[len(buffer)-1]
	actualCRC := CRC8XOR(data)

	return actualCRC == expectedCRC
}

File: admin/stdlib/crc.c
#include "crc.h"

uint8_t crc8_xor(const uint8_t* data, size_t len) {
    uint8_t crc = 0;
    for (size_t i = 0; i < len; ++i) {
        crc ^= data[i];
    }

    return crc;
}
